### PR TITLE
Fix ballerina-shell unicode not identifying issue

### LINKED ISSUE #30021: Enable disabled code sample in shell tests
**Description:**
Enable disabled shell test code sample with https://github.com/ballerina-platform/ballerina-lang/pull/29977/commits/81b275a553d35d7b2ab1cf4d46313daa13c1d214 in PR #29977.

Sample:
```
  {
    "description": "Use unicode with escape in name.",
    "code": "string 'unicode_\\u{2324} = \"Jane doe\"; 'unicode_\\u{2324}",
    "expr": "\"Jane doe\""
  }
```
It gives,
```
java.lang.AssertionError: Exception occurred in: Use unicode with escape in name., error: Invocation of the snippet failed., with diagnostics: [Node identified as a VARIABLE_DECLARATION snippet., Node identified as a EXPRESSION snippet., error: undefined symbol 'unicode_'
	'unicode_⌤ // Variable is unable to be referenced after declaration
	^-------^, Compilation aborted due to errors.]
```

### PR DESCRIPTION
**Pull request by DamithaSenevirathne**
## Purpose

Fixes #30021 

## Approach

Issue occurred due to java Unicode type is different from the ballerina Unicode type. Added a function to convert Unicode types.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/StringUtils.java [hunk 1]
```diff
@@ -162,4 +162,19 @@ public static String getErrorStringValue(Throwable error) {
...
         return error.getMessage();
     }
+
+    /**
+     * Convert unicode character to ballerina unicode format
+     *
+     * @param string Input unicode containing string to convert.
+     * @return reformatted unicode string.
+     */
+    public static String convertUnicode(String string){
```
Comment (kdsuneraavinash) [inline ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/StringUtils.java:172]: Maybe simplify as following?

```java
    public static String convertUnicode(String string){
        int character = string.charAt(0);
        return "\\" + "u" + "{" + Integer.toHexString(character) + "}";
    }
```

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/StringUtils.java [hunk 2]
```diff
@@ -137,7 +137,7 @@ public static String unescapeUnicodeCodepoints(String identifier) {
      * @return encoded string.
      */
     public static String encodeIdentifier(String string) {
-        return IdentifierUtils.encodeNonFunctionIdentifier(string);
+        return string.replace("\\", "\\\\");
```
Comment (kdsuneraavinash) [inline ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/StringUtils.java:140]: Will this replacement be enough? I think we may have to test some other cases involving quotes, etc...

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/QuotedIdentifier.java [hunk 3]
```diff
@@ -27,11 +27,24 @@
  * @since 2.0.0
  */
 public class QuotedIdentifier {
-    private final String name;
+    private final int ASCII_VALUE = 128;
```
Comment (kdsuneraavinash) [inline ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/QuotedIdentifier.java:30]: Can we make this a static variable?
Also can we change the name, this is more so of a limit, no?

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/QuotedIdentifier.java [hunk 4]
```diff
@@ -27,11 +27,24 @@
...
+    private String name;
```
Comment (kdsuneraavinash) [inline ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/QuotedIdentifier.java:31]: Can we keep this as a final variable?

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/QuotedIdentifier.java [hunk 5]
```diff
@@ -27,11 +27,24 @@
...
     public QuotedIdentifier(String name) {
+        boolean isUnicode = false;
         Objects.requireNonNull(name);
-        this.name = StringUtils.quoted(name);
+        char[] characters = name.toCharArray();
+        for (char character: characters) {
+            if (character >= ASCII_VALUE) {
+                isUnicode = true;
+                String string = "" + character;
+                this.name = StringUtils.quoted(name)
```
Comment (kdsuneraavinash) [inline ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/QuotedIdentifier.java:41]: I think there is a problem here, `this.name` gets updated in the loop using `name`, so changes done in previous iterations are discarded.

I think its better to do this as following, (`convertUnicode` method may need to be changed to accept a character)
```java
    public QuotedIdentifier(String name) {
        Objects.requireNonNull(name);
        StringBuilder convertedName = new StringBuilder();
        for (char character : name.toCharArray()) {
            if (character >= ASCII_VALUE) {
                String replacement = StringUtils.convertUnicode(character);
                convertedName.append(replacement);
            } else {
                convertedName.append(character);
            }
        }
        this.name = StringUtils.quoted(convertedName.toString());
    }
```

**Review by kdsuneraavinash — CHANGES_REQUESTED** [review]
I think the test case was also not re-enabled.

**Comment by DamithaSenevirathne** [comment]
Test case enabled in 93dc4c5911efb39793c57a9d21af9878c06df06d

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/StringUtils.java (hunk 1 above)
Comment (DamithaSenevirathne) [reply ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/StringUtils.java]: Updated in 3083628406bc711786e3b472071f89c46d21e09e

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/QuotedIdentifier.java (hunk 3 above)
Comment (DamithaSenevirathne) [reply ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/QuotedIdentifier.java]: Updated in 3083628406bc711786e3b472071f89c46d21e09e

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/QuotedIdentifier.java (hunk 4 above)
Comment (DamithaSenevirathne) [reply ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/QuotedIdentifier.java]: Updated in 3083628406bc711786e3b472071f89c46d21e09e

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/QuotedIdentifier.java (hunk 5 above)
Comment (DamithaSenevirathne) [reply ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/QuotedIdentifier.java]: Updated in 3083628406bc711786e3b472071f89c46d21e09e

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/StringUtils.java (hunk 2 above)
Comment (DamithaSenevirathne) [reply ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/StringUtils.java:140]: It seems that this replacement is enough for the test cases that we are currently using (including quotes)

**Review by kdsuneraavinash — APPROVED** [review]

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/StringUtils.java [hunk 6]
```diff
@@ -164,17 +164,12 @@ public static String getErrorStringValue(Throwable error) {
     }
 
     /**
-     * Convert unicode character to ballerina unicode format
+     * Convert unicode character to ballerina unicode format.
      *
      * @param string Input unicode containing string to convert.
```
Comment (kdsuneraavinash) [inline ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/StringUtils.java]: Update doc comment

**Review by kdsuneraavinash — CHANGES_REQUESTED** [review]

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/StringUtils.java (hunk 6 above)
Comment (DamithaSenevirathne) [reply ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/StringUtils.java]: Updated in bd20d609da6b7d469096e76b8fad2d9ccdeb2cf9

**Review by kdsuneraavinash — APPROVED** [review]

File: ballerina-shell/modules/shell-core/src/test/resources/testcases/evaluator/basics.quoted.json [hunk 7]
```diff
@@ -59,5 +59,10 @@
     "description": "Member access of escaped name.",
     "code": "person2[\"first name\"]",
     "expr": "\"Tom\""
+  },
+  {
+    "description": "Use unicode with escape in name.",
+    "code": "string 'unicode_\\u{2324} = \"Jane doe\"; 'unicode_\\u{2324}",
+    "expr": "\"Jane doe\""
```
Comment (pubudu91) [inline ballerina-shell/modules/shell-core/src/test/resources/testcases/evaluator/basics.quoted.json:66]: Shall we add some of the other cases you tested as well?

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/StringUtils.java [hunk 8]
```diff
@@ -162,4 +162,14 @@ public static String getErrorStringValue(Throwable error) {
         }
         return error.getMessage();
     }
+
+    /**
+     * Convert unicode character to ballerina unicode format.
+     *
+     * @param character unicode character to convert.
+     * @return reformatted unicode string.
```
Comment (pubudu91) [inline ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/StringUtils.java:170]: Use sentence case for these documentation as well.

**Comment by pubudu91** [comment]
Shall we sync this PR with master again?

File: ballerina-shell/modules/shell-core/src/test/resources/testcases/evaluator/basics.quoted.json (hunk 7 above)
Comment (DamithaSenevirathne) [reply ballerina-shell/modules/shell-core/src/test/resources/testcases/evaluator/basics.quoted.json:66]: Added in 0ce501dc6d16f3464a59ae4f5636ac25167baa84

File: ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/StringUtils.java (hunk 8 above)
Comment (DamithaSenevirathne) [reply ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/utils/StringUtils.java]: Updated in 2b06d029352f8a27893b7d41e10e3ad95aa851ea

**Review by kdsuneraavinash — APPROVED** [review]

**Pull request merged**